### PR TITLE
Pings

### LINKED ISSUE #78: After losing connection client is in invalid state
This problem is similar to #73  .
After losing connection to NATS Streaming server client is in invalid state. It can't close and/or dispose connection (instance of `STAN.Client.IStanConnection`), in this case application just stops working, waiting for connection to close. If one doesn't close the connection, application can't close as in #73 case.
Code to reproduce error with NATS Streaming server deployed on localhost:

```
class Program
{
  static void Main(string[] args)
  {
    var opts = StanOptions.GetDefaultOptions();
    opts.NatsURL = "127.0.0.1:4222";
    var cf = new StanConnectionFactory();
    var conn = cf.CreateConnection("test-cluster", "TEST_CLIENT_SENDER", opts);
    conn.Subscribe("test1", (obj, msgArgs) =>
    {
                    Console.Write($"[{System.Threading.Thread.CurrentThread.ManagedThreadId}] Message from subscription \"{msgArgs.Message.Subscription}\": ");
                    Console.WriteLine($"{System.Text.Encoding.UTF8.GetString(msgArgs.Message.Data)} \"");
    });
    
    while (true)
    {
      string inp = Console.ReadLine();
      if (inp is null || inp == "q") break;
      try
      {
        conn.Publish("test1", System.Text.Encoding.UTF8.GetBytes(msg));
      }
      catch
      {
        //If connection is lost, first StanException will occur (because of timeout),
        //then we try to close old connection and release resources. Yet we fail.
        //To simulate loss of connection, one can simply restart NATS Streaming server.
        //Leave one of two lines below
        conn.Close();
        //conn.Dispose();
        conn = cf.CreateConnection("test-cluster", "TEST_CLIENT_SENDER", opts);
        conn.Publish("test1", System.Text.Encoding.UTF8.GetBytes(msg));
      }
    }
  }
}
```

I found a workaround: instead of closing connection in `STAN.Client.IStanConnection` one can call `STAN.Client.IStanConnection.NATSConnection.Close()` and client will be no longer in invalid state. But there's still an error: when one is closing connection that way and connecting again to NATS Streaming there is still memory leak, something around 1.5 MB.

### PR DESCRIPTION
**Pull request by ColinSullivan1**
This is the C# client implementation of:  https://github.com/nats-io/go-nats-streaming/pull/183

This will help the client library detect and report that
the connection to the Streaming server is permanently lost.

This also allows the server to detect that a client has
been replaced by another one with the same client ID.
Imagine a client being disconnected from the streaming
server. At the same time, another application that can
communicate with the streaming server creates a connection
with the same client ID. The server detects the duplicate
ID, tries to contact the original client, but can't,
assume that it has terminated and replaces it.
When the original client resumes communication with the
server, it should now fail because the connection is
no longer "valid". For that, the client PINGs contain
a connection ID that uniquely identify any connection (NUID).
When the server gets a PING it verifies if that connection
ID is still valid, if not send back an error.

If the user has registered a new ConnectionLostHandler
(with the option setter SetConnectionLostHandler), he/she
will be notified of the error.

Also, published messages now contain this connection ID, which
allows the server to reject a published message if the
client is no longer valid.

The NATS connection created by the library during the Connect()
will now set the maximum reconnect attempts to infinite and
the reconnect buffer size to -1 to prevent any buffering during
a disconnect event.

The library uses default Ping interval of 5seconds and a max
number of PINGs without response of 3. These values can be
changed by the addition of a new option setter: Pings().

NOTE: No ping is used when connecting to an older server.
Also, if a client originally connects to a new server and that
new server is shutdown and an older server is restarted, the
ConnectionLostHandler will be invoked since the old server
does not listen to client PINGs.

Resolves #78

### CODE REVIEW

File: STAN.CLIENT/StanConnection.cs [hunk 1]
```diff
@@ -234,6 +267,207 @@ internal Connection(string stanClusterID, string clientID, StanOptions options)
...
+
+                    lock (pingLock)
+                    {
+                        pingTimer = new Timer(pingServer, null, pingInterval, Timeout.Infinite);
+                    }
+                }
+            }
+            if (unsubPing)
+            {
+                pingSubscription.Unsubscribe();
```
Comment (watfordgnf) [inline STAN.CLIENT/StanConnection.cs]: it looks like a few things above (e.g. ConnectRequest work) can cause an exception to be throw after pingSubscription is created on line 202.

File: STAN.CLIENT/Options.cs [hunk 2]
```diff
@@ -150,6 +156,51 @@ public long MaxPubAcksInFlight
...
+
+                pingInterval = value;
+            }
+        }
+
+        /// <summary>
+        /// Represents the method that will handle an event raised
+        /// when a connection has been disconnected from a streaming server.
+        /// </summary>
+        public EventHandler<StanConnLostHandlerArgs> ConnectionLostEventHandler = null;
```
[the remark this answers is not in the document]
Comment (ColinSullivan1) [reply STAN.CLIENT/Options.cs]: Thanks - will fix...

File: STAN.CLIENT/StanConnection.cs [hunk 3]
```diff
@@ -234,6 +267,207 @@ internal Connection(string stanClusterID, string clientID, StanOptions options)
...
+        // Everytime a PING is sent, the number of outstanding PINGs is increased.
+        // If the total number is > than the PingMaxOut option, then the connection
+        // is closed, and connection error callback invoked if one was specified.
+        private void pingServer(object state)
+        {
+            IConnection conn = null;
+            Exception pingEx = null;
+            bool lostConnection = false;
+
+            // we're closed, just exit
```
[the remark this answers is not in the document]
Comment (ColinSullivan1) [reply STAN.CLIENT/StanConnection.cs]: Thanks.

File: STAN.CLIENT/Options.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ColinSullivan1) [reply STAN.CLIENT/Options.cs]: Good catch - Thanks.

File: STAN.Client.UnitTests/UnitTestBasic.cs [hunk 4]
```diff
@@ -1811,5 +1811,305 @@ public void TestSubscriberClose()
...
+                    so.NatsURL = "nats://127.0.0.1:4333";
+                    so.PingInterval = 1000;
+                    so.PingMaxOutstanding = 120;
+                    so.ConnectionLostEventHandler = (obj, args) =>
+                    {
+                        errStr = args.ConnectionException.Message;
+                        ev.Set();
+                    };
+
+                    sc1 = scf.CreateConnection(CLUSTER_ID, CLIENT_ID, so);
```
[the remark this answers is not in the document]
Comment (ColinSullivan1) [reply STAN.Client.UnitTests/UnitTestBasic.cs]: Yeah, I should make a pass through the tests for that.

File: STAN.CLIENT/StanConnection.cs [hunk 5]
```diff
@@ -234,6 +267,207 @@ internal Connection(string stanClusterID, string clientID, StanOptions options)
...
+            if (data?.Length > 0)
+            {
+                var pingResp = new PingResponse();
+                try
+                {
+                    ProtocolSerializer.unmarshal(data, pingResp);
+                }
+                catch
+                {
+                    return;
```
[the remark this answers is not in the document]
Comment (ColinSullivan1) [reply STAN.CLIENT/StanConnection.cs]: It'll be ignored as a bad ping, but this certainly deserves a comment.

File: STAN.CLIENT/StanConnection.cs [hunk 6]
```diff
@@ -234,6 +267,207 @@ internal Connection(string stanClusterID, string clientID, StanOptions options)
...
+
+            // Unsubscribe only if we have a connection...
+            if (nc != null && !nc.IsClosed()) {
+                ackSubscription?.Unsubscribe();
+                pingSubscription?.Unsubscribe();
+            }
+
+            // fail all pending pubs...
+            PublishAck pa;
+            var keys = pubAckMap.Keys;
```
[the remark this answers is not in the document]
Comment (ColinSullivan1) [reply STAN.CLIENT/StanConnection.cs]: Appreciate the attention to detail - any access to PubAckMap should be under the connection lock `mu`, so is protected.   I'll add a function comment about the caller locking under `mu`.

**Comment by ColinSullivan1** [comment]
All tests pass for .NET 4.5, Core seems to get hung on one test when AppVeyor is very busy, haven't been able to get past that today.  IMO, we should merge.  Wdyt?

**Comment by dungpa** [comment]
@ColinSullivan1 @watfordgnf Another vote for merging this PR. Thanks.

**Comment by watfordgnf** [comment]
Agree on squash.

**Pull request merged**